api/aggregation.proto

```diff
@@ -0,0 +1,181 @@
+syntax = "proto3";
+
+package aggregation;
+
+import "validate/validate.proto";
+
+// [#next-free-field: 3]
+message KeyerConfiguration {
+
+  // [#next-free-field: 2]
+  message Fragment {
+
+    // A rule defining how to match a Envoy request and what resulting
+    // fragment to generate.
+    // [#next-free-field: 3]
+    message Rule {
+
+      // Defines how to match an Envoy Request.
+      MatchPredicate match = 1 [(validate.rules).message.required = true];
+
+      // Defines how to generate the resulting fragment if matched.
+      ResultPredicate result = 2 [(validate.rules).message.required = true];
+    }
+
+    // The first rule that matches will be applied in sequential order.
+    repeated Rule rules = 1 [(validate.rules).repeated .min_items = 1];
+  }
+
+  // Separator for each key fragment. Default if not specified is "_".
+  string fragments_separator = 1;
+
+  // Fragments are the pieces that form a cache key.
+  repeated Fragment fragments = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+enum NodeFieldType {
+  NODE_ID = 0;
+  NODE_CLUSTER = 1;
+  NODE_LOCALITY_REGION = 2;
+  NODE_LOCALITY_ZONE = 3;
+  NODE_LOCALITY_SUBZONE = 4;
+}
+
+// This is a recursive structure which allows complex nested match
+// configurations to be built using various logical operators.
+// [#next-free-field: 7]
+message MatchPredicate {
+
+  // Rules for matching on a Envoy request type.
+  // [#next-free-field: 2]
+  message RequestTypeMatch {
+    // One or more Envoy request types to match on.
+    // Ex: "type.googleapis.com/envoy.api.v2.Listener"
+    repeated string types = 1 [(validate.rules).repeated .min_items = 1];
+  }
+
+  // Match on a Field in Envoy's request node.
+  // [#next-free-field: 4]
+  message RequestNodeMatch {
+
+    // Specifies the Envoy Request field that should be matched on.
+    NodeFieldType field = 1 [(validate.rules).enum = {defined_only: true}, (validate.rules).message.required = true];
+
+    oneof type {
+      option (validate.required) = true;
+
+      // Match on the extract string value.
+      string exact_match = 2;
+
+      // Match on a regex pattern.
+      // TODO potentially use "safe regex"
+      // https://github.com/envoyproxy/envoy/blob/10f756efa17e56c8d4d1033be7b4286410db4e01/api/envoy/type/matcher/v3/regex.proto
+      string regex_match = 3;
```

Comment:
Why can't w use regex now?

Reply:
No reason, I was just copying the proto definition from @jessicayuen 's spec, but I think we can commit to use envoy's regex matcher from the get go.

Reply:
Could you investigate into what that would look like @eapolinario? It should be relatively simple, the main thing is to ensure that the constraints set by safe regex don't limit what we may want to match on for a request node field (ex: wildcard prefixes).

Reply:
We talked offline about this and we reached the conclusion that instead of relying on envoy's regex matcher type we should define our own, however, this PR is already gigantic as is, so... how about we tackle this in follow up PR?